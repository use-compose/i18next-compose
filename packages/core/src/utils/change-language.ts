import { useLogger } from '@use-compose/logger';
import { i18n } from 'i18next';

export const localStorageLangKey = 'lang';

export async function changeLanguage(i18n: i18n, lang: string) {
  const { log } = useLogger('changeLanguage');

  i18n.changeLanguage(lang, (err, t) => {
    if (err) return log('something went wrong loading - ' + err);
    t('key'); // -> same as i18next.t
  });

  if (typeof window !== 'undefined') {
    localStorage.setItem(localStorageLangKey, lang);
  }
}
