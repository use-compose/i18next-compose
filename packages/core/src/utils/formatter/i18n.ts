import { i18n, Namespace, TOptions } from 'i18next';
import type { I18nFormatterHelper } from '../../types';

export const i18NextFormatterMock = (): I18nFormatterHelper => {
  const translationHelper = () => '';
  const globalNSHelper = () => '';

  return { translationHelper, globalNSHelper };
};

export type ComposeI18nHelper = (level3: Namespace | string, params?: TOptions) => string;

/**
 * Description
 * @param i18nextConfig: i18n
 * @returns {I18nFormatterHelper}
 
 */
export function i18nFormatter(i18nextConfig: i18n): I18nFormatterHelper {
  if (!i18nextConfig) {
    return i18NextFormatterMock();
  }

  /**
   * Create and return a translation helper based on the current namespace
   * @param level2: string
   * @returns {ttFunc}
   */
  const translationHelper = (level2: string): ComposeI18nHelper => {
    // t contains the translation function from the given configuration

    const t = i18nextConfig.t;
    /*
     * We do not enforce type of ns, since we need to also dynamically access plurals
     * e.g. we do not want to write _one or _other in the namespace
     */
    function composeT(level3: Namespace | string, params?: TOptions): string {
      return t(`${level2}.${level3}`, params) ?? '';
    }

    // composeT.namespace = level2;
    return composeT;
  };

  /**
   * Return helper associated to global namespace
   * @param ns
   */
  const globalNSHelper = (level3: string): string => {
    return translationHelper('global')(level3);
  };

  return { translationHelper, globalNSHelper };
}
