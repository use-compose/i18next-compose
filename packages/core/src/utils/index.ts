export { changeLanguage } from './change-language';
export type { ChangeLanguageFunc } from './change-language';
export { i18nFormatterHelper, i18nFormatterMock } from './formatter/translation-helpers';
export { getFlagEmoji } from './iso-to-emoji';
