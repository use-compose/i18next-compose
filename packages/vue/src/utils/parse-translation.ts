export {
  componentsPresent,
  hasManyChildren,
  isLowercaseHtmlTag,
  parseTranslation,
  removeNumberSuffix,
  type FlatComponent,
};

/**
 * TAG_NAME is used to build the regexes
 * - It matches tag names like:
 *   - PascalCase: <NuxtLink>
 *   - kebab-case: <my-component> or HTML tags like <strong>
 *   - Component with numeric suffixes: <NuxtLink-1>, <my-component-2>
 */
const TAG_NAME = '([A-Za-z][\\w-]*)'; // allows PascalCase, kebab-case, and suffixes like -1

const i18nRegex = {
  /*
   * We need to consider different cases:
   * 1 Opening tags (e.g. <NuxtLink>)
   * 2 Closing tags (e.g. </NuxtLink>)
   * 3 Self-closing tags (e.g. <NuxtLink />, which in this case could refer to a class as the logo in the login page)
   * 4 Tag with optional numeric suffix (e.g. <NuxtLink-1> or <NuxtLink-2>) if this component is used multiple times in the same translation string
   *
   * We will then inject the props directly in the template that call the  ̶R̶t̶T̶r̶a̶n̶s̶l̶a̶t̶e̶ Translate component
   * in the case 3 it's more straightforward but in the case 1 and 2 we need to also keep the content inside the tag
   *
   * These are 3 RegEx that we can use to detect components or tags in the translation value
   * All of them also support snake-case in case of third-party components
   *
   */

  TAG_NAME: '([A-Za-z][\\w-]*)', // allows PascalCase, kebab-case, and suffixes like -1

  // selfClosingAndOpeningTagsOnly: new RegExp(`<${TAG_NAME}\\s*/>`, 'g'),

  /**
   * Matches self-closing tags
   * E.g. <Tag ... />
   * Or in the case of multiple same tags with suffixes: <Tag-1 ... />
   *
   * Groups:
   *  1: tag name
   *  2: attributes (unused)
   */
  selfClosingTags: new RegExp(`<${TAG_NAME}\\s*([^>]*)\\s*/>`, 'g'),

  // selfClosingOpeningAndClosingTags: new RegExp(
  //   `<${TAG_NAME}\\s*([^>]*)\\s*(?:/)?>(?:</${TAG_NAME}\\s*>)?|</${TAG_NAME}\\s*>`,
  //   'g',
  // ),

  /**
   * Matches all types of tags: opening, closing, and self-closing
   * E.g. <Tag ...>, </Tag>, <Tag ... />
   * Or in the case of multiple same tags with suffixes: <Tag-1 ...>, </Tag-1>, <Tag-1 ... />
   *
   * Used for recursive parsing when we need to find all tags in a string
   *
   * Groups:
   *  1: opening/self-closing tag name
   *  2: attributes (unused)
   *  3: self-closing slash (if any)
   *  4: closing tag name
   */
  allTypesOfTags: new RegExp(`<${TAG_NAME}\\s*([^>]*)\\s*(/?)>|</${TAG_NAME}\\s*>`, 'g'),

  // Paired OR self-closing, with content captured when paired.
  // Groups:
  //  1: paired tag name, 2: attributes (unused), 3: inner content
  //  4: self-closing tag name
  allTagsWithContent: new RegExp(
    `<${TAG_NAME}\\s*([^>]*)\\s*(?:/)?>([\\s\\S]*?)</\\1\\s*>|<${TAG_NAME}\\s*/>`,
    'g',
  ),

  /**
   * Matches opening and closing tags, and captures inner content.
   * E.g. <Tag ...>inner</Tag>
   * Or in the case of multiple same tags with suffixes: <Tag-1 ...>inner</Tag-1>
   *
   * Groups:
   *  1: tag name
   *  2: attributes (unused)
   *  3: inner content
   */
  openingAndClosingTagsWithContent: new RegExp(
    `<${TAG_NAME}\\s*([^>]*)\\s*(?:/)?>([\\s\\S]*?)</\\1\\s*>`,
    'g',
  ),

  standaloneTags: new RegExp(`<${TAG_NAME}\\s*([^>]*)\\s*(?:/)?>(?:</\\1\\s*>)?`, 'g'),
};

/**
 * Types
 */
export interface TagObject {
  tag: string;
  content?: TagObject[] | string;
}

interface FlatComponent {
  tag: string;
  content: string;
}

export type ParsedResult = (string | TagObject)[]; // | ParsedResult[]

/**
 * Utility to check if a parsed tag object has multiple children
 *
 * @param {TagObject} element
 * @returns {boolean}
 */
function hasManyChildren(element: TagObject): boolean {
  if (!element.content) {
    return false;
  }
  return Array.isArray(element.content);
}

/*
 * Used to handle cases when a translation string contains multiple tags with the same name
 * e.g. "<NuxtLink>Wilkommen</NuxtLink> <NuxtLink-2>zurück</NuxtLink-2>""
 * We remove the hyphen and the number to only render the real tag name
 */
function removeNumberSuffix(str: string) {
  return str.replace(/-\d+$/, ''); // only at the end
}

/*
 * Will return an array of strings of tags or null
 * Example: ["<NuxtLink>","</NuxtLink>"]
 */
function componentsPresent(translationString: string): string[] | null {
  return translationString.match(i18nRegex.allTypesOfTags);
}

function isLowercaseHtmlTag(name: string) {
  return /^[a-z]/.test(name);
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function isSelfClosingTag(str: string): boolean {
  const selfClosingTags = str.match(i18nRegex.selfClosingTags);
  return selfClosingTags ? selfClosingTags.includes(str) : false;
}

/*
 * Parse a given translation string and return an array of strings and objects
 * e.g. for a key
 * "Wunderbar! <NuxtLink>The customer <strong>{{ customer.name }}</strong></NuxtLink> hat den <NuxtLink-2>Kurs</NuxtLink-2> erfolgreich absolviert.",
 *
 * Will return
 * [
 *   'Wunderbar! ',
 *   {
 *     tag: 'NuxtLink',
 *     content: [
 *       'The customer ',
 *       {
 *         tag: 'strong',
 *         content: 'Arthur',
 *       },
 *     ],
 *   },
 *   ' hat den ',
 *   {
 *     tag: 'NuxtLink-2',
 *     content: 'Kurs',
 *   },
 *   ' erfolgreich absolviert.',
 * ]
 */

function parseTranslation(translationString: string): ParsedResult {
  const result: ParsedResult = [];

  // Explicit, readable patterns (support kebab-case and -1 suffixes)
  const PAIRED_RE = i18nRegex.openingAndClosingTagsWithContent;
  const SELF_RE = i18nRegex.selfClosingTags;

  let cursor = 0;

  // Find the next earliest tag (paired or self-closing) from a given position
  function findNext(from: number) {
    PAIRED_RE.lastIndex = from;
    SELF_RE.lastIndex = from;

    const paired = PAIRED_RE.exec(translationString);
    const self = SELF_RE.exec(translationString);

    // choose whichever starts earlier; if equal, prefer paired (so <X>…</X> wins over <X/> at same spot)
    const pick =
      paired && self
        ? paired.index <= self.index
          ? { kind: 'paired', m: paired }
          : { kind: 'self', m: self }
        : paired
          ? { kind: 'paired', m: paired }
          : self
            ? { kind: 'self', m: self }
            : null;

    if (!pick) return null;

    const start = pick.m.index;
    const end = (pick.kind === 'paired' ? PAIRED_RE : SELF_RE).lastIndex;

    if (pick.kind === 'paired') {
      const tagName = pick.m[1]; // name for paired
      const inner = pick.m[3] ?? ''; // inner content
      return { kind: 'paired' as const, tagName, inner, start, end };
    } else {
      const tagName = pick.m[1]; // name for self-closing
      return { kind: 'self' as const, tagName, start, end };
    }
  }

  while (cursor < translationString.length) {
    const next = findNext(cursor);
    if (!next) break;

    // push plain text before this tag
    if (next.start > cursor) {
      result.push(translationString.slice(cursor, next.start));
    }

    if (next.kind === 'self') {
      result.push({ tag: next.tagName });
    } else {
      // paired tag
      if (next.inner && componentsPresent(next.inner)) {
        result.push({
          tag: next.tagName,
          content: parseTranslation(next.inner) as string | TagObject[],
        });
      } else {
        result.push({ tag: next.tagName, content: (next.inner ?? '').trim() });
      }
    }

    cursor = next.end;
  }

  // trailing text
  if (cursor < translationString.length) {
    result.push(translationString.slice(cursor));
  }

  return result;
}
