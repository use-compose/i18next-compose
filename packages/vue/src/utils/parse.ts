export {
  componentsPresent,
  hasManyChildren,
  parseTranslation,
  removeNumberSuffix,
  type FlatComponent,
};

const i18nRegex = {
  /*
   * We need to consider 3 cases:
   * 1 Opening tags (e.g. <NuxtLink>)
   * 2 Closing tags (e.g. </NuxtLink>)
   * 3 Self-closing tags (e.g. <NuxtLink />, which in this case could refer to a class as the logo in the login page)
   *
   * We will then inject the props directly in the template that call the RtTranslate component
   * in the case 3 it's more straightforward but in the case 1 and 2 we need to also keep the content inside the tag
   *
   * These are 3 RegEx that we can use to detect components or tags in the translation value
   * All of them also support snake-case in case of third-party components
   */

  // 1 & 3 - could be considered to inject the props
  // https://regex101.com/r/Or6Bah/1
  selfClosingAndOpeningTagsOnly: /<([A-Z][a-zA-Z0-9]*|[a-z][a-z0-9-]*)\s?\/?>/g,

  // 1 & 2 & 3
  // https://regex101.com/r/axPbVS/1
  selfClosingOpeningAndClosingTags:
    /<([A-Z][a-zA-Z0-9]*|[a-z][a-z0-9-]*)\s*([^>]*)\s*(?:\/)?>|<\/([A-Z][a-zA-Z0-9]*|[a-z][a-z0-9-]*)\s*>/g,

  // Same but slightly different, to test
  // https://regex101.com/r/iZVQDh/1
  allTypesOfTags:
    /<([A-Z][a-zA-Z0-9]*|[a-z][a-z0-9-]*)\s*([^>]*)\s*(\/?)>|<\/([A-Z][a-zA-Z0-9]*|[a-z][a-z0-9-]*)\s*>/g,
  /*
   * "The first regular expression uses a non-capturing group for the optional slash,
   * while the second regular expression directly captures the optional slash." (from ChqtGPT)
   */

  /*
   * RegEx that also contain the content inside the tags
   */
  // https://regex101.com/r/TQXaEL/1
  allTagsWithContent:
    /<([A-Z][a-zA-Z0-9]*|[a-z][a-z0-9-]*)\s*([^>]*)\s*(?:\/)?>((?:.|[\r\n])*?)<\/\1\s*>|<([A-Z][a-zA-Z0-9]*|[a-z][a-z0-9-]*)\s*([^>]*)\s*(?:\/)?>/g,

  /*
   * RegEx to match tags and content inside it
   */
  // https://regex101.com/r/3SXgzD/1
  openingAndClosingTagsWithContent:
    /<([A-Z][a-zA-Z0-9-]*|[a-z][a-z0-9-]*)\s*([^>]*)\s*(?:\/)?>((?:.|[\r\n])*?)<\/\1\s*>/g,

  // https://regex101.com/r/w9BBjg/1
  standaloneTags: /<([A-Z][a-zA-Z0-9]*|[a-z][a-z0-9-]*)\s*([^>]*)\s*(?:\/)?>/g,
};

export interface TagObject {
  tag: string;
  content?: TagObject[] | string;
}

interface FlatComponent {
  tag: string;
  content: string;
}

export type ParsedResult = (string | TagObject)[]; // | ParsedResult[]

function hasManyChildren(element: TagObject): boolean {
  if (!element.content) {
    return false;
  }
  return Array.isArray(element.content);
}

/*
 * Used to handle cases when a translation string contains multiple tags with the same name
 * e.g. "<NuxtLink>Wilkommen</NuxtLink> <NuxtLink-2>zurück</NuxtLink-2>""
 * We remove the hyphen and the number to only render the real tag name
 */
function removeNumberSuffix(str: string) {
  return str.replace(/-\d+/g, '');
}

/*
 * Will return an array of strings of tags or null
 * Example: ["<NuxtLink>","</NuxtLink>"]
 */
function componentsPresent(translationString: string): string[] | null {
  return translationString.match(i18nRegex.allTypesOfTags);
}

/*
 * Parse a given translation string and return an array of strings and objects
 * e.g. for a key
 * "Wunderbar! <NuxtLink>The customer <strong>{{ customer.name }}</strong></NuxtLink> hat den <NuxtLink-2>Kurs</NuxtLink-2> erfolgreich absolviert.",
 *
 * Will return
 * [
 *   'Wunderbar! ',
 *   {
 *     tag: 'NuxtLink',
 *     content: [
 *       'The customer ',
 *       {
 *         tag: 'strong',
 *         content: 'Arthur',
 *       },
 *     ],
 *   },
 *   ' hat den ',
 *   {
 *     tag: 'NuxtLink-2',
 *     content: 'Kurs',
 *   },
 *   ' erfolgreich absolviert.',
 * ]
 */
function parseTranslation(translationString: string): ParsedResult {
  let initialValue = translationString;

  const result: ParsedResult = [];

  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace
  initialValue.replace(
    i18nRegex.openingAndClosingTagsWithContent,
    function (substring, tagName, _attrs, content) {
      const index = initialValue.indexOf(substring);

      /*
       * If there is a string before the tag, we push it to the result
       * e.g. "Hallo und <NuxtLink>Wilkommen zurück</NuxtLink>! Viel Spaß"
       * We push "Hallo und " to the result
       */
      if (index > 0) {
        result.push(initialValue.substring(0, index));
      }

      if (componentsPresent(content)) {
        // If the content of the tag contains other tags, we parse it recursively
        content = parseTranslation(content);
        result.push({ tag: tagName, content });
      } else {
        result.push({ tag: tagName, content: content.trim() });
      }

      initialValue = initialValue.substring(index + substring.length);
      /**
       * TODO: to avoid
       * Argument of type '(substring: string, tagName: any, _attrs: any, content: any) => void' is not assignable to parameter of type '(substring: string, ...args: any[]) => string'.
       * Type 'void' is not assignable to type 'string'.
       */
      return substring;
    },
  );

  // TODO:
  // Match self-closing tags
  // initialValue.replace(i18nRegex.standaloneTags, function (match, tagName) {
  //   const index = initialValue.indexOf(match)
  //   if (index > 0) {
  //     result.push(initialValue.substring(0, index))
  //   }
  //   result.push({ tag: tagName })
  //   initialValue = initialValue.substring(index + match.length)
  // })

  // Push the remaining string if any
  if (initialValue.length > 0) {
    result.push(initialValue);
  }

  return result;
}
