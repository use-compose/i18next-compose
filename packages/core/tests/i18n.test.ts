import { i18n } from 'i18next';
import { beforeEach, describe, expect, test } from 'vitest';
import { changeLanguage } from '../src/utils/change-language';
import { getFormatterFunctions, initi18nConfig } from './i18n';

describe('Create a custom config and access translation keys through formatter', () => {
  let i18next: i18n;

  beforeEach(async () => {
    i18next = await initi18nConfig();
  });

  test('i18next configuration', async () => {
    expect(i18next).toBeDefined();
    expect(i18next.language).toEqual('en');
    expect(i18next.languages).toEqual(['en']);
  });

  test('Access translation keys through formatter', async () => {
    const { translationHelper, globalNSHelper } = getFormatterFunctions(i18next);
    expect(translationHelper).toBeDefined();
    expect(globalNSHelper).toBeDefined();

    const cT = translationHelper('compose_translations');
    expect(cT).toBeDefined();

    const keyValue = cT('test_key');
    expect(keyValue).toEqual('Value of test key2');
    expect(cT('test_key')).toMatchSnapshot();
  });

  test('Change language and access translation keys', async () => {
    await changeLanguage(i18next, 'de');
    expect(i18next.language).toEqual('de');
    expect(i18next.languages).toEqual(['de', 'en']);
    const { translationHelper } = getFormatterFunctions(i18next);
    const cT = translationHelper('compose_translations');
    expect(cT).toBeDefined();
    expect(cT('test_key')).toEqual('Wert des Testschlüssels2');
    expect(cT('test_key')).toMatchSnapshot();
  });

  test('Dynamically change language by accessing a key in different language', async () => {
    const { translationHelper, globalNSHelper } = getFormatterFunctions(i18next);
    expect(translationHelper).toBeDefined();
    expect(globalNSHelper).toBeDefined();

    const cT = translationHelper('compose_translations');
    expect(cT).toBeDefined();

    const keyValue = cT('test_key');
    expect(keyValue).toEqual('Value of test key2');

    await changeLanguage(i18next, 'de');
    expect(i18next.language).toEqual('de');
    expect(i18next.languages).toEqual(['de', 'en']);

    const keyValueDe = cT('test_key');
    expect(keyValueDe).toEqual('Wert des Testschlüssels2');
    expect(cT('test_key')).toMatchSnapshot();
  });
});
