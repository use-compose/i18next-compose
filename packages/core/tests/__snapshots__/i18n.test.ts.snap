// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Create a custom config and access translation keys through formatter > Access translation keys through formatter 1`] = `"Value of test key2"`;

exports[`Create a custom config and access translation keys through formatter > Change language and access translation keys 1`] = `"Wert des Testschlüssels2"`;

exports[`Create a custom config and access translation keys through formatter > Dynamically change language by accessing a key in different language 1`] = `"Wert des Testschlüssels2"`;
